Normalize Database

tas09009 opened the pull request:
### Changes to the schema
- remove Nicety: `starred` and User: `autosave_timeout` and `autosave_enabled` This will address Issues #80 and #34 
- Connected `User` to `Profile` using `User.profile_id`
- Nicety: `target_id` now points to `Profile.id`. Added back `name` along with other attributes that were normally being called thorough the API.
- Stint: Added in `profile_id` for the foreign key connection
- Profile: removed `faculty` since that's determined by `stint['type']` 

#### To use new database:
- `createdb <database_name>`
- change the `DATABASE_URL` environment variable in your `.env` file to update the database, then run `source .env` on your terminal
- `flask db upgrade`
- Run `python update-data.py`on the project level directory. The script logs each `Profile` being added so feel free to comment out lines 78 and 83.

Here is a link to [edit the schema directly](https://drive.google.com/file/d/1OzvtIwecEpZbpSPcJ7R0NBynYQyFcJbo/view?usp=sharing)
![image](https://user-images.githubusercontent.com/8211605/133318410-a976b6ea-4cdb-4b52-9d5e-1806f99f6046.png)

jasonaowen commented on migrations/versions/12780a039923_populate_database.py:
```diff
@@ -0,0 +1,82 @@
+    sa.Column('interests', sa.String(), nullable=True),
+    sa.Column('before_rc', sa.String(), nullable=True),
+    sa.Column('during_rc', sa.String(), nullable=True),
+    sa.PrimaryKeyConstraint('id')
```
Please indent these parenthetical blocks!

jasonaowen commented on migrations/versions/12780a039923_populate_database.py:
```diff
@@ -0,0 +1,82 @@
+    )
+    op.add_column('nicety', sa.Column('stint', sa.Integer(), nullable=True))
+    op.drop_constraint('nicety_author_id_target_id_end_date_key', 'nicety', type_='unique')
+    op.create_unique_constraint(None, 'nicety', ['author_id', 'target_id'])
```
This constraint needs to include the stint, so that one person can write two niceties for someone who has attended RC twice.

jasonaowen commented on migrations/versions/12780a039923_populate_database.py:
```diff
@@ -0,0 +1,82 @@
+depends_on = None
+
+
+def upgrade():
```
We need to migrate existing data here, including finding existing `User`s - who may no longer be part of RC, and thus not be in the data returned by the RC API - and populating the new `profile` table with that data.

jasonaowen commented on migrations/versions/12780a039923_populate_database.py:
```diff
@@ -0,0 +1,82 @@
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_table('profile',
+    sa.Column('id', sa.Integer(), nullable=False),
+    sa.Column('name', sa.String(length=500), nullable=True),
+    sa.Column('first_name', sa.String(), nullable=True),
+    sa.Column('last_name', sa.String(), nullable=True),
+    sa.Column('avatar_url', sa.String(length=500), nullable=True),
+    sa.Column('bio_rendered', sa.String(), nullable=True),
+    sa.Column('interests', sa.String(), nullable=True),
+    sa.Column('before_rc', sa.String(), nullable=True),
+    sa.Column('during_rc', sa.String(), nullable=True),
```
Definitely some of these should be nullable, but I'd be curious if we can tighten any of these up. In the real data, do we always have `avatar_url`? Or `name`?

tas09009 replied:
Your're right, `name` should not be nullable. This ended up being the only consistent attribute between `user` and `profile` (the old `user` table doesn't have first and last names) so that is the only nullable attribute I found.

jasonaowen replied:
Awesome, thanks for checking!

jasonaowen commented on migrations/versions/12780a039923_populate_database.py:
```diff
@@ -0,0 +1,82 @@
+    )
+    op.create_table('stint',
+    sa.Column('id', sa.Integer(), nullable=False),
+    sa.Column('profile_id', sa.Integer(), nullable=True),
+    sa.Column('type_stint', sa.String(), nullable=True),
+    sa.Column('start_date', sa.Date(), nullable=True),
+    sa.Column('end_date', sa.Date(), nullable=True),
```
I think all of these are required, and should not be nullable.

tas09009 replied:
I made these nullable, thanks for catching this. The only exception I found was `end_date` since faculty don't have any end date.

jasonaowen reviewed (changes requested):
Awesome work, @tas09009! Initial comments on the database structure below. You should now have access to the prod data, so please download a copy and try this migration locally.

jasonaowen commented on update-data.py:
```diff
@@ -0,0 +1,116 @@
+            db.session.commit()
+
+        else:
+            logging.info(f"Skipping: {id}")
```
I think we do need to update existing profiles! Particularly with the initial `User` migration, but also as people do things like change their interests or upload new photos.

tas09009 replied:
Made changes to this so it adds/updates every profile, thank you!

jasonaowen commented on migrations/versions/12780a039923_populate_database.py:
```diff
@@ -0,0 +1,82 @@
+    # ### end Alembic commands ###
+
+
+def downgrade():
```
I think there's going to end up being quite a lot more to this downgrade than is here, in order to handle existing data. I also don't think that's a worthwhile use of time - we'll take a database backup before merging this PR, and if the deployment goes wrong, we can roll back and manually restore. We can just drop the `downgrade` entirely!

tas09009 commented:
Originally we talked about creating one migration script which includes the schema updates as well as the new models to populate from the RC API. Here is actually the approach I took (so far):

In order to normalize the database with `profile` and `stint`, I needed to 
1. create and migrate those models
2. run the database populating script
3. create another migration to move over missing people from `user` to `profile`.

I couldn't combine steps 1 and 3 into one migration script since the database has to be populated with profiles *after* the `profile` and `stint` tables are created but *before* the missing profiles from `user` can be added in. Here are the steps to run them on the production database.
1. `flask db downgrade` run this command *twice* to downgrade to the latest migration script used in production
2. `flask db upgrade b99a55eb8e08` to add `Profile` and `Stint` tables
3. `python update-data.py`to add the 2070 profiles
4. `flask db upgrade 27d86222146d` to migrate missing users (not pulled from RC API) over to `Profile`. Total profiles should now be 2074. 

If we don't want two migrations, then I imagine a single migration can include: 
- creating `profile` and `stint` tables and moving all `user`s over to `profile` so that `profile` now has 451 `user`s.
- running the `update-data.py` to pull all RC API profiles and overwriting the 451 ids that already exist in `profile` but not overwriting the missing ones.

I also ran into a separate issue: apparently there several niceties written for `target_id=1932` but `id=1932` doesn't exist in `user` or `profile`. Which means I'm unable to change the schema to connect the `nicety.target_id` to `profile.id`. I don't think this is the only case since one or two other ids came up like this yesterday. Not sure why this is? Until then, I can't update the schema for `nicety`.

I still made changes to the `user` and `nicety` models without adding them to any migration. If this is confusing, I can change them back to their original state until they are ready to be migrated.

Let me know if this is unclear or the wrong approach to take. Thanks!

tas09009 commented:
Just asked James about those missing ids and he said they are no longer part of the RC community. I can speak more to this at our Wednesday meeting.

tas09009 commented:
I believe this captures everything we've talked about thus far. Run the following lines on the terminal to upgrade and downgrade the migration:

`flask db upgrade`
`flask db downgrade`

jasonaowen commented:
@tas09009 and I paired on restructuring these changes. So far we've pulled two commits out, 448a9c3 and e0b4921.

e0b4921 is worth calling out, I think: I reviewed this data-only migration by examining the database by hand.

Nicety ID 221 is one of the lost niceties described in #10: the target user changed their end date, and the author created a new nicety (ID 791) that is very similar but not identical, and with the new end date.

Relevant queries:

```sql
SELECT id, author_id, target_id, date_updated, CONVERT_FROM(DECODE(text, 'base64'), 'utf8')
FROM nicety WHERE id IN (221, 791);
SELECT * FROM nicety WHERE author_id = ? AND target_id = ?;
```

Nicety ID 3634 is more complicated. It has a correct end date, but it is the replacement for a previously-written nicety (ID 2956) that was lost due to the target recurser changing their end date. The text of the old nicety has much more detail than the new one, but it also has an incorrect end date. I believe the intent is to clean that up in the subsequent migration that assigns `stint_id`s to all niceties that have an end date that matches no known stints. Also, @tas09009, is this one of the niceties people you emailed the author about?

tas09009 commented:
Hi @jasonaowen, I didn't send an email for this one because of what you mentioned when comparing the two niceties. But I certainly can if you'd like!

tas09009 committed:
Add update_data script and update README

"update_data.py" fetches information from the RC API to populate
the database and have an up-to-date cache. We can schedule data updates
using the Heroku Scheduler.

Update the "README.md" to include instructions on creating an access
token for the RC API requests and set up a Heroku Scheduler.

Issue #68: Normalize Recurser profiles
Issue #5: Clearing the cache in a timely fashion

Co-authored-by: Jason Owen <jason@jasonaowen.net>

tas09009 committed:
Update previous migrations to match current ones

Remove unused comments, add a downgrade, and change the formatting
to match new migrations.

tas09009 committed:
Remove unused attributes from Nicety and User

Remove nicety.starred, user.autosave_enabled and user.autosave_timeout.
These columns are not used and haven't been for some time.

Issue #34: Remove starred column
Issue #80: Remove unused autosaved columns

tas09009 committed:
Remove user.anonymous_by_default column

Giving niceties anonymously in bulk is a partially implemented
feature for the user.

Ultimately want to discourage this feature as well as make it easier for
code maintenance.

Issue #79: Remove unused anonymous_by_default column

tas09009 committed:
Delete two duplicate niceties

These two niceties were manually reviewed and had substantially similar
but not identical content to two other niceties that are from the same
author_id to the same target_id.

These niceties were "lost" because the target user extended their batch
and had a different end_date.

Issue #82: Normalize Recurser Profiles
Issue #10: Niceties lost when Recursers extend their batch

tas09009 committed:
Create Profile and Stint tables

We want to maintain a local copy of the RC API data and make our data
model more consistent by including all the relevant information for a
nicety.

Previously, we were relied on making API calls to populate
data we didn't have locally in response to each request to the front
end. We are still doing this, but these data models set the stage to
stop doing that.

Issue #68: Normalize Recurser profiles

Co-authored-by: Jason Owen <jason@jasonaowen.net>

tas09009 committed:
Populate Profile and Stint tables

Create new tables and models for Profiles and Stints, and populate
them with the data from the RC API.

Import functions from update_data.py. Following this
migration, use the update_data.py module to pull all data from
the RC API.

Co-authored-by: Jason Owen <jason@jasonaowen.net>

tas09009 committed:
Create Profiles for missing people

Create Profiles for:

- 5 people: ids missing from the Profile table but found in
  the User table.
- 4 people: ids missing from both User and Profile tables.
  These 4 people have niceties written for them but no profile
  to connect to. Save their names as "Former Recurser".

Co-authored-by: Jason Owen <jason@jasonaowen.net>

tas09009 committed:
Add nicety.stint_id column + attributes

The nicety.stint_id column will remain nullable until all profiles
are populated with their respective stints in future migrations.

Create a foreign key constraint to tie niceties to stints. Change the
unique constraint to be between the "author_id," "target_id," and
"stint_id." We are still leaving "end_date" alone because we are not
yet cleaning up the backend to use "stints" instead of "end_dates."

Create a foreign key to tie target_ids to profile_ids.

Co-authored-by: Jason Owen <jason@jasonaowen.net>

tas09009 committed:
Populate nicety.stint_ids using nicety.end_dates

Populate all nicety.stint_ids based on end dates since it's
a required column. This change will exclude 156 niceties due to
mismatching end dates.

The subsequent migration will address these 156 niceties with null
stint_ids.

tas09009 committed:
Create Stints for missing people

The 156 niceties with null stint_ids consist of 18 profiles with 18
end dates.

Out of the 18 profiles, create stints for the newly created profiles
using negative stint ids:

- 9 new profiles manually created (4 Former Recurser's + 5 from
  Users table)
- 2 profiles with no niceties written for them, therefore no need to
  generate stints

This leaves 7 profiles with 8 stints, with 2 stints for one profile

After this, there will be 68 niceties with null stint_ids left.
The subsequent migration will address this.

tas09009 committed:
Populate nicety.stints ids with the latest stint

Populate stint_ids for 68 niceties where nicety.stint_id = null.
All niceties are for the target_id's latest stint.

tas09009 committed:
Change nicety.stint_id to never be nullable

Nicety.stint_id was allowed to be null while populating their values.

Now that the 156 nicety.stint_ids are populated, this column won't be
null again.

tas09009 committed:
Update README.md RC application link

Update the external link to set up an RC application and restrict the
text width to be less than 85 columns wide.

Co-authored-by: Jason Owen <jason@jasonaowen.net>

jasonaowen commented on migrations/versions/689a6ce963c3_remove_anonymous_columns.py:
```diff
@@ -0,0 +1,33 @@
+    code maintenance.
+    """
+    op.drop_column("user", "anonymous_by_default")
+    op.drop_column("nicety", "anonymous")
```
I think we do still want to allow anonymous niceties! Just not saving the default preference.

jasonaowen commented on migrations/versions/b49314aac7b0_create_stints_for_missing_users.py:
```diff
@@ -0,0 +1,103 @@
+    connection = op.get_bind()
+
+    """
+    The 156 nicetys with null dates can be grouped into 18 profiles
```
I think this is 156 niceties with null *stint_id*, rather than null dates, right?

jasonaowen commented:
So last we talked, the migrations were in pretty good shape, and the remaining work on this PR is to update the API to use the new database structure. Thanks again for working on this, @tas09009 - I know it's a big change!

tas09009 committed:
API Changes (WIP)

These are the changes to the actual backend files to finally pull
data from the database, rather than user RC's API.

This is a draft; the files will need to be organized and old code
will need to be deleted. This push is simply to have the changes
on Github.

tas09009 commented:
Hi @jasonaowen,

While working on the API changes, I realized they could only happen *after* the database migration since some of the changes we're implementing - switching from batches to stints and removing caching - have to wait until that last migration. You can see some of my WIP in the 3rd to last [commit](https://github.com/mjec/rc-niceties/pull/82/commits/5424b413bfd4075d53c3fc8324135da298bb3319).

Do you think we should split this PR up again into two parts? It'll be easier to digest too:
1. Database + schema migration
2. API changes to incorporate migrations

It will also make it easier to push changes to the two PRs as unanticipated code changes come up, such as when we discover that a user's stint changes their `date` and their `stint_id`.
